include/merlin_hashtable.cuh

```diff
@@ -272,84 +309,97 @@ class HashTable {
       if (metas == nullptr) {
         upsert_kernel_with_io<key_type, vector_type, meta_type, DIM, TILE_SIZE>
             <<<grid_size, block_size, 0, stream>>>(
-                table_, keys, reinterpret_cast<const vector_type*>(values),
-                table_->buckets, table_->buckets_size, table_->bucket_max_size,
+                table_, keys, value_vectors, table_->buckets,
+                table_->buckets_size, table_->bucket_max_size,
                 table_->buckets_num, N);
       } else {
         upsert_kernel_with_io<key_type, vector_type, meta_type, DIM, TILE_SIZE>
             <<<grid_size, block_size, 0, stream>>>(
-                table_, keys, reinterpret_cast<const vector_type*>(values),
-                metas, table_->buckets, table_->buckets_size,
-                table_->bucket_max_size, table_->buckets_num, N);
+                table_, keys, value_vectors, metas, table_->buckets,
+                table_->buckets_size, table_->bucket_max_size,
+                table_->buckets_num, N);
       }
     } else {
-      vector_type** d_dst = nullptr;
-      int* d_src_offset = nullptr;
+      auto device_ws = device_memory_pool_->get_workspace<2>(stream);
```

Comment:
if the `options_.io_by_cpu` is not `true`, can we avoid allocating the buffer?

Reply:
This is a device buffer... You need that to store the offsets. The `find` function doesn't have a `options_.io_by_cpu` execution path yet.